### PR TITLE
[test/plugin]  immigrate  undertow scenario

### PR DESCRIPTION
**Pull request by arugal**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues
#3583
___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

**Comment by dmsolr** [comment]
I am not sure, maybe these classes are not enhanced by `undertow-plugin` under `org.apache.skywalking.*` package.

File: Jenkinsfile-Agent-Test [hunk 1]
```diff
@@ -72,57 +72,57 @@ pipeline {
...
-                        stage('jetty 9.x (63)') {
-                            steps {
-                                sh 'bash test/plugin/run.sh --build_id=wl1_${BUILD_ID} jetty-scenario'
-                            }
-                        }
-                    }
-                }
-            }
-        }
+//        stage('Run Agent Plugin Tests') {
```
Comment (wu-sheng) [inline Jenkinsfile-Agent-Test]: Why this?

**Comment by wu-sheng** [comment]
Are you pushing local change to PR? You commented all other test cases. And please move the test to workload 1 or 2. Workload 3 is the slowest one today.

**Comment by arugal** [comment]
@dmsolr @wu-sheng please review

File: test/plugin/scenarios/undertow-scenario/support-version.list [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+2.0.15.Final
+2.0.16.Final
+2.0.17.Final
+2.0.19.Final
+2.0.20.Final
+2.0.21.Final
+2.0.22.Final
+2.0.23.Final
+2.0.26.Final
+2.0.27.Final
```
Comment (wu-sheng) [inline test/plugin/scenarios/undertow-scenario/support-version.list]: Could you add some comments and descriptions here about why we removed some versions to be tested? And how we select these versions?

File: test/plugin/scenarios/undertow-routing-scenario/support-version.list [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+2.0.15.Final
+2.0.16.Final
+2.0.17.Final
+2.0.19.Final
+2.0.20.Final
+2.0.21.Final
+2.0.22.Final
+2.0.23.Final
+2.0.26.Final
+2.0.27.Final
```
Comment (wu-sheng) [inline test/plugin/scenarios/undertow-routing-scenario/support-version.list]: Same version selection rule should be doc here.

File: test/plugin/scenarios/undertow-scenario/support-version.list (hunk 2 above)
Comment (arugal) [reply test/plugin/scenarios/undertow-scenario/support-version.list]: done

**Comment by wu-sheng** [comment]
@arugal Next time, please add test cases into workload 1 and 2. Remind again. The workload 3 is slowest already.

**Comment by wu-sheng** [comment]
@dmsolr Please review.

**Comment by wu-sheng** [comment]
FYI @kezhenxu94, as this PR is only plugin test related, e2e skips as expected.

**Comment by arugal** [comment]
> Is the old version like this? I notice, the ref does not exist. Should have a httpclient calls undertow service, which make ref existing. Could you enhance this test?

done

**Comment by arugal** [comment]
> @arugal Next time, please add test cases into workload 1 and 2. Remind again. The workload 3 is slowest already.

I have moved ```undertow-scenario``` to workload2

File: test/plugin/scenarios/undertow-routing-scenario/src/main/java/org/apache/skywalking/amp/testcase/undertow/Application.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        };
+        RoutingHandler handler = new RoutingHandler();
+        handler.add(Methods.GET, TEMPLATE, httpHandler);
+        handler.add(Methods.HEAD, TEMPLATE, httpHandler);
+        Undertow server = Undertow.builder()
+            .addHttpListener(8080, "127.0.0.1")
+            .setHandler(handler).build();
+        Runtime.getRuntime().addShutdownHook(new Thread(server::stop));
+        // Waiting for service register, please do not delete.
+        Thread.sleep(5000);
```
Comment (dmsolr) [inline test/plugin/scenarios/undertow-routing-scenario/src/main/java/org/apache/skywalking/amp/testcase/undertow/Application.java]: Here is waiting for Agent Register done. Do you have any other suggestions about this, @wu-sheng @kezhenxu94 .

Comment (wu-sheng) [reply test/plugin/scenarios/undertow-routing-scenario/src/main/java/org/apache/skywalking/amp/testcase/undertow/Application.java]: Provide a test API toolkit and plugin to interact with agent?

Comment (dmsolr) [reply test/plugin/scenarios/undertow-routing-scenario/src/main/java/org/apache/skywalking/amp/testcase/undertow/Application.java]: sounds like a great idea.

Comment (wu-sheng) [reply test/plugin/scenarios/undertow-routing-scenario/src/main/java/org/apache/skywalking/amp/testcase/undertow/Application.java]: @dmsolr Do you know how to do it? Do you need my help?

Comment (arugal) [reply test/plugin/scenarios/undertow-routing-scenario/src/main/java/org/apache/skywalking/amp/testcase/undertow/Application.java]: Ask ```mock-collect``` before health check to determine whether the agent is registered?

Comment (arugal) [reply test/plugin/scenarios/undertow-routing-scenario/src/main/java/org/apache/skywalking/amp/testcase/undertow/Application.java]: But such cases should be rare, and ```undertow``` is a special case

Comment (dmsolr) [reply test/plugin/scenarios/undertow-routing-scenario/src/main/java/org/apache/skywalking/amp/testcase/undertow/Application.java]: Yes. as @arugal say, I am preparing to push the PR.

**Comment by wu-sheng** [comment]
You break the test case.

**Comment by arugal** [comment]
Please do not merge this PR yet, I also have an enhance to the plugin

**Comment by wu-sheng** [comment]
Ping us after you done.

**Comment by arugal** [comment]
@wu-sheng @dmsolr Hi, I added the ```io.undertow.server.HttpServerExchange#dispatch(Executor, Runnable)``` intercept, please review.

**Comment by wu-sheng** [comment]
> Hi, I added the io.undertow.server.HttpServerExchange#dispatch(Executor, Runnable) intercept, please review.

Is this a missing intercept point? What is the case we need this, and existing interceptors are not supported? Could you give some introduce?

**Comment by arugal** [comment]
> Is this a missing intercept point? 

Yes

> What is the case we need this, and existing interceptors are not supported?

As shown below, in ```HttpHandler#handleRequest(HttpServerExchange)```, an asynchronous task can be submitted through ```HttpServerExchange#dispatch(Executor, Runnable)```
 and this interception was previously missing.
```
        Undertow server = Undertow.builder()
            .addHttpListener(8080, "0.0.0.0")
            .setHandler(exchange -> {
                exchange.dispatch(new HttpHandler() {
                    @Override
                    public void handleRequest(HttpServerExchange httpServerExchange) throws Exception {
                        // do something async
                    }
                });
                exchange.dispatch(new Runnable() {
                    @Override
                    public void run() {
                        // do something  async
                    }
                });
                exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "text/plain");
                exchange.getResponseSender().send("Success");
            }).build();
        server.start();
```

**Pull request merged**